Testing: Check + fix 'import from' statements

dgmcdona opened the pull request:
Updates the AST code analysis script to detect `from volatitity3... import ...` statements for imports of non-module objects, and fixes all existing noncompliant code. Also updates the script name to reflect the fact that it is more generally for code analysis instead of specifically for checking requirements.

@ikelos Going ahead and opening this as targeting the branch with the original work, but will re-target once #1743 is merged.

dgmcdona committed:
Testing: Adds validation of vol3 imports in check script

This checks `ast.ImportFrom` statements to see if anything other than
modules are being imported in this way. It enumerates all instances of
this and suggests a fix.

dgmcdona committed:
Framework: Fix all direct non-module imports

This fixes all import from statements in the codebase that were
importing things other than modules into module namespaces from other
volatility3 modules. This should prevent accidental re-exporting.

dgmcdona committed:
CI Testing: Renames script and updates job name

ikelos commented:
Looks like there's some registry circular imports.  I'd check for some of them just being needed for type checking, in which case they can usually be wrapped in quotes which delays the need for the class until it's all been loaded...

dgmcdona committed:
Comment type annotation to fix circular import

dgmcdona commented on volatility3/framework/interfaces/configuration.py:
```diff
@@ -37,7 +37,8 @@
 )
 
-from volatility3 import classproperty, framework
+import volatility3
```
This suggestion would make for some very long type hints, I'd rather not. Should we leave things as-is or maybe make an exception for `classproperty` in the checker?

ikelos replied:
Yeah, those type hints are nuts (there's zero reason to be use the full name, even if we wanted to be that long winded).  It also looks like it's trying to subclass `HierarchicalDict` from itself?!?!  They're one in the same class?

I'm happy to give `abstract.abc` a free pass (like we do with `typing`), so then it would be `from collections.abc import ABCMeta, abstractmethod, Mapping` and then `HierarchicalDict` can subclass `Mapping`.  Hopefully that will keep the suggestion bot at bay, otherwise I'm happy to dismiss this concern...

ikelos replied:
I'm not sure why this patch imports `volatility3` and then `volatility3.framework` though?  I don't think they're needed I would definitely let classproperty be a direct import too.

ikelos replied:
So maybe just 
```
from volatility3 import classproperty
from volatility3.framework import constants, interfaces
```
?

dgmcdona committed:
Code Analysis: Give pass to 'volatility3'

Also moves some code into a private method with a docstring in the
visitor class.

dgmcdona committed:
Revert changes to configuration.py

dgmcdona committed:
Revert changes to intel.py

dgmcdona committed:
Make `registry_layers` -> `registry_layer` for consistency

ikelos reviewed (approved):
Thanks, this looks really good!
